custom_ops2.py

```diff
@@ -0,0 +1,208 @@
+# custom op API #2: BackendImpl and Function.
+#
+# In this version of the API, there are two main custom op APIs:
+# - Use library.BackendImpl to define a black-box operation with an abstract impl
+# - Use library.Function to define transform rules (e.g. autograd formula) for
+# a sequence of make_fx-able operations (e.g. 1+ BackendImpl, existing
+# PyTorch operators, or triton kernels)
+#
+# This approach is more explicit about traceability (Function's
+# implementation MUST be traceable) and also allows automagic schema
+# inference (because Function is always traceable, we can infer a schema
+# for it).
+# =====================================================================
+
+# =====================================================================
+# To use PyTorch (and torch.compile) with a third-party library or
+# custom kernel, create a BackendImpl.
+#
+# Morally, this represents some low-level kernel and its abstract impl.
+# PyTorch (e.g. torch.compile) will NEVER peek into BackendImpl -- it
+# is always treated as a black-box.
+#
+# BackendImpl is not differentiable (or transformable, e.g. with functorch)
+# by itself: to add an autograd formula or another transform rule, see below.
+class NumpySin(BackendImpl):
+    # Maybe we can infer this schema from the first call :P
+    schema = "(Tensor x) -> Tensor"
+
+    # the black-box cpu kernel
+    @staticmethod
+    def impl_cpu(x):
+        result = torch.from_numpy(np.sin(x.detach().cpu().numpy()))
+        return result
+
+    # the black-box cuda kernel
+    @staticmethod
+    def impl_cuda(x):
+        return torch.from_numpy(np.sin(x.detach().cpu().numpy())).to(x.device)
+
+    # the abstract impl. Must be "traceable". User must use opcheck to test.
+    @staticmethod
+    def abstract(x):
+        return torch.empty_like(x)
+
+
+# =====================================================================
+# To specify custom functionality for a subsystem like autograd or vmap,
+# a user should use Function (torch.library.Function).
+#
+# Functions should be functionally pure, e.g. their outputs are only
+# determined from the inputs are there are no observable side effects.
+#
+# Under the hood, `Function` automatically constructs 1+ custom ops and
+# uses Dispatch keys to add functionality.
+#
+# We require all staticmethods of `Function` be make_fx traceable:
+# that is, torch.compile will peek into them. All `Function` will decompose
+# into their forward() staticmethod when they hit Inductor.
+#
+# This is intentionally confusingly named like torch.autograd.Function,
+# because it is the thing you should use instead of torch.autograd.Function :).
+class MySin(Function):
+    @staticmethod
+    def forward(x):
+        return NumpySin.apply(x)
+
+    @staticmethod
+    def post_forward(ctx, args, output):
+        x, = args
+        ctx.save_for_backward(x)
+
+    @staticmethod
+    def backward(ctx, grad_out):
+        return grad_out * ctx.saved_tensors[0].cos()
```

Comment:
cc @tugsbayasgalan it's plausible that we could just take an autograd.Function and turn it into a custom op, with or without Dynamo, under the assumption that it is make_fx-traceable (and well behaved). The `.apply` on this automatically creates 1+ opoverloads depending on what the input types are.